### PR TITLE
Visca protocol

### PR DESCRIPTION
**Pull request by Davidsastresas**
Support for cameras using visca protocol, like the popular sony fcb block cameras. update rate in the scheduler has been set to 3 hz. By the moment it supports a 3 position switch for zoom+, zoom- and stop zoom.

For configuring in autopilot, select serialx protocol to 18, and baud rate according to the camera's one.
select VSC_CTRL_CH with the channel number of the switch desired for zoom and VSC_CTRL_MODE = 0 (default).

I plan to add in the next days support for more operation modes, and more useful mesages as day/night mode IR filter on/off, digital zoom enabled/disabled.... As well as update the wiki for the users to be able to use it.

This is my first contribution to ardupilot, I've tried to make everything as required in the wiki and according to the rest of the code I've seen, but please let me know if I need to change anything to fit in the existing code.

Regards.

### CODE REVIEW

**Comment by Davidsastresas** [comment]
I've just read the script of the devcall of yesterday, sorry I couldn't assist.

I totally agree with the ideas you said there. In fact I thought on integrating it into AP_Mount at first, but I thought maybe you wouldn't agree to it. In fact I am also  "polishing" a bit the alexmos protocol, because I am working on a video tracker based on raspberry pi 3, and I would like to integrate it all as a afordable solution for a "serious" payload on ardupilot. I used to work on something like that on a private company and I would like to share it to the open source as there is nothing similar right now.

I also don't like the use of a whole new k_parameter because of this, I would be much neater inside AP_mount, and use AP_mount as a "full payload"

I also thought on doing it inside the AP_Camera, but as it is now it is mostly for mapping, so that's why I thought on a new library.

In any case, let me know the final decision about this and I will happily commit the necessary changes. 

Thanks,

regards.

**Comment by tridge** [comment]
thanks for getting back to us. I think moving it into AP_Mount as a new backend type is the right way to go. If you don't mind re-working it that way that would be appreciated

**Comment by Davidsastresas** [comment]
Sure, right now I am on vacations, but as soon as I get back I'll work on it. Thanks.

Regards.

**Comment by Davidsastresas** [comment]
@tridge , I am again back from vacations and working on this, but I have to ways of aproaching it.

If I define a new AP_Mount_Backend_Camera base class, and I inherit a particular Visca class from this one, which would be the neater solution, I need to modify a bit the AP_Mount class, because right now AP_Mount uses pointers to AP_Backend for working with all the different kinds of mounts (Alexmos, stm32...) and if I define a new Backend for cameras, I need to implement in AP_Mount this new pointer to AP_Mount_Backend_Camera for handling this. 

The other approach I have in mind, is make the Visca class inheriting from AP_Mount_Backend, as this class has most of the virtual methods I need for impement visca. The drawback of this, is that AP_Mount_Backend also has lot of methods refering to angles and such that I won't be using on Visca, and overall wouldn't be as neat as a new backend just for cameras.

Please tell me which one of them you consider more appropriate, as I am not as familiarized with ardupilot for considering which one of them would be best, regarding memory usage and such.

In both cases I would need to add a few parameters on the AP_Mount group, I hope it is all right.

Thank you.

Regards.

**Comment by tridge** [comment]
I like the 2nd approach, making AP_Mount_Visca inherit from AP_Mount_Backend. If when you do it you find it gets all too messy then jump on mumble to discuss with one of the core devs

**Comment by tridge** [comment]
> In both cases I would need to add a few parameters on the AP_Mount group, I hope it is all right.

note that there is a way to have params for a particular backend. See the AP_BattMonitor for an example

**Comment by Davidsastresas** [comment]
Didn't know that about particular params for each backend, I will take a look. Thank you very much, I'll work on it.

Regards.

File: ArduCopter/ArduCopter.cpp [hunk 1]
```diff
@@ -143,6 +143,9 @@ const AP_Scheduler::Task Copter::scheduler_tasks[] = {
 #if CAMERA == ENABLED
     SCHED_TASK_CLASS(AP_Camera,            &copter.camera,              update_trigger,  50,  75),
 #endif
+#if VISCA == ENABLED
```
Comment (amilcarlucas) [inline ArduCopter/ArduCopter.cpp]: Replace `VISCA` with `VISCA_ENABLED`
Add '//#define VISCA_ENABLED DISABLED` in the APM_Copter.h file

**Comment by Davidsastresas** [comment]
I am sorry, I've been on other face of my project related to computer vision this last weeks. Just let me a few days and I will  implement the changes requested by Tridge. 

Thank you,

regards.

File: ArduCopter/ArduCopter.cpp (hunk 1 above)
Comment (Davidsastresas) [reply ArduCopter/ArduCopter.cpp]: I updated that already, but I don't know how to mark as solved this particular changes you requested. I am working on the rest of the PR, I hope to get it on the beginning of the next week.

Thank you.

**Comment by Davidsastresas** [comment]
I've made a clean in the commits history because it was getting messy with all the undos. Right now has the same functionality as before all this changes, just supporting direct control zoom, +, - and stop zoom. I plan to add support for progressive zoom ( proportional to channel value ) and mavlink mode, with new mavlink messages and all, but by the moment is functional.

I've moved it as another mount, inheriting from AP_Mount_Backend as suggested. The changes to the AP_Mount.cpp and .h have been minimal, so I think it is quite clean. 

I looked at the suggestion of @tridge of parameters for each backend as batt monitor, but, correct me if I am wrong please, that would require a bigger change in AP_Mount, as for making it clean I think would require to implement that idea of parameters for each backend for each current backend, not just for Visca, so probably the cleanest thing to do it would be to add another Mount_Parameters class or something like that. 

Right now It just needed 2 new parameters, and I don't thing it is gonna need much more. I've written this parameters under MNT2, so in the regular code, with MAX_INSTANCES = 1, the two visca parameters are not used. In any case, using this kind of high end zoom camera doesn't make sense without a gimbaled payload, so I think it will always work as secondary Mount.

I didn't implemented the changes requested by @amilcarlucas as right now all the code is contained within AP_Mount, it is no longer under arducopter files so I wasn't quite sure of where to put the "IF VISCA_ENABLED == DISABLED". Right now, it could work putting the VISCA_ENABLED == DISABLED on Mount.h, most code related with visca would not be compiled but there would be a small aditions under SerialManager that will remain, just the default baudrates and the enum kinds of serial and that. I wanted to ask you first about how to put this, if you consider all right to put VISCA_ENABLED == DISABLED on AP_Mount.h I will make the necessary changes.

Also if you consider something should be different please let me know and I will work on it, I have still little experience on Ardupilot and hopefully after a few more contributions I will need less guidance, but I would appreciate all the suggestions you can made.

Thank you.

File: libraries/AP_Mount/AP_Mount.h [hunk 2]
```diff
@@ -28,6 +28,7 @@
 #include <AP_SerialManager/AP_SerialManager.h>
 #include <DataFlash/DataFlash.h>
 
+
```
Comment (amilcarlucas) [inline libraries/AP_Mount/AP_Mount.h]: Do not add blank lines

File: libraries/AP_Mount/AP_Mount.h [hunk 3]
```diff
@@ -73,7 +77,8 @@ class AP_Mount
         Mount_Type_SoloGimbal = 2,      /// Solo's gimbal
         Mount_Type_Alexmos = 3,         /// Alexmos mount
         Mount_Type_SToRM32 = 4,         /// SToRM32 mount using MAVLink protocol
-        Mount_Type_SToRM32_serial = 5   /// SToRM32 mount using custom serial protocol
+        Mount_Type_SToRM32_serial = 5,   /// SToRM32 mount using custom serial protocol
```
Comment (amilcarlucas) [inline libraries/AP_Mount/AP_Mount.h]: One whitespace too much.
Please add a Visca comment

Comment (Davidsastresas) [reply libraries/AP_Mount/AP_Mount.h]: I'm sorry, I should have noticed that before pushing. Fixed.

File: libraries/AP_Mount/AP_Mount.h (hunk 2 above)
Comment (Davidsastresas) [reply libraries/AP_Mount/AP_Mount.h]: Sorry, should have checked before pushing. Fixed

**Comment by davidbuzz** [comment]
needs rebase and no action since 2018.

**Comment by Davidsastresas** [comment]
If there is still interest in this I can re do it over updated master. Please confirm me it is of interest for the project and I will rebase

**Comment by Shaikfpv** [comment]
Hello David, iam still interested in this and would like to have a firmware with visca control for matek f405 ctr flight controller. Can you please help me with it? Thanks in advance

**Comment by Davidsastresas** [comment]
I no longer have access to visca equipment, and I don't know how soon I will be able to test the rebase of this, so I think it is better to close it for now. Thanks and sorry for the trouble.

**Pull request closed without merging**